### PR TITLE
languages/kotlin: init

### PR DESCRIPTION
**Pull request by Soliprem**
# Kotlin Support

Enables Kotlin LSP/TS support

## Type of change

- New feature (non-breaking change which adds functionality)

### Notes
Startup time is slower than normal, but it seems to be known. I added the init option to speed it up, and picked what I think is a sensible default. This doesn't impact functionality in any way other than startup time.

## Checklist

Please try to check at least a majority of the checklist before opening your pull request. Exceptions to this will be reviewed on a case by case basis.

- [X] My code follows the style and contributing guidelines of this project.
- [X] I ran Alejandra to format my code (`nix fmt`).
- [X] I have performed a self-review of my own code and tested it.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] My changes generate no new warnings.
- [ ] This change requires a documentation update.
- [ ] I have updated the documentation accordingly.

## Screenshots & Logs

image

### CODE REVIEW

File: modules/plugins/languages/kotlin.nix [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+  cfg = config.vim.languages.kotlin;
+
+  # Creating a version of the LSP with access to the kotlin binary.
+  # This is necessary for the LSP to load the standard library
+  kotlinLspWithRuntime = pkgs.symlinkJoin {
+    name = "kotlin-language-server-with-runtime";
+    paths = [
+      pkgs.kotlin-language-server
+      pkgs.kotlin
```
Comment (NotAShelf) [inline modules/plugins/languages/kotlin.nix:22]: Is kotlin-language-server not wrapped with kotlin?

File: modules/plugins/languages/kotlin.nix [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    ];
+    buildInputs = [pkgs.makeWrapper];
+    postBuild = ''
+      wrapProgram $out/bin/kotlin-language-server \
+        --prefix PATH : ${pkgs.lib.makeBinPath [pkgs.kotlin]}
+    '';
+  };
+in {
+  options.vim.languages.kotlin = {
+    enable = mkEnableOption "kotlin/HCL support";
```
Comment (NotAShelf) [inline modules/plugins/languages/kotlin.nix:32]: Please capitalise the K in Kotlin here and in other instances.

File: modules/plugins/languages/kotlin.nix [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    treesitter = {
+      enable = mkEnableOption "kotlin treesitter" // {default = config.vim.languages.enableTreesitter;};
+      package = mkGrammarOption pkgs "kotlin";
+    };
+
+    lsp = {
+      enable = mkEnableOption "kotlin LSP support (kotlin_language_server)" // {default = config.vim.languages.enableLSP;};
```
Comment (NotAShelf) [inline modules/plugins/languages/kotlin.nix:40]: In theory, language modules allow alternating LSP servers, so hardcoding the name of a single server here is redundant.

**Review by NotAShelf — CHANGES_REQUESTED** [review]
Looks good overall. Few changes requested for consistency sake.

In addition, it would be nice to see formatter and diagnostics support if possible - I see none-ls supports [Ktlint](https://pinterest.github.io/ktlint/latest/).

**Review by NotAShelf — CHANGES_REQUESTED** [review]
Looks good overall. Few changes requested for consistency sake.

In addition, it would be nice to see formatter and diagnostics support if possible - I see none-ls supports [Ktlint](https://pinterest.github.io/ktlint/latest/).

File: modules/plugins/languages/kotlin.nix (hunk 1 above)
Comment (Soliprem) [reply modules/plugins/languages/kotlin.nix:22]: Not that I'm aware of

**Comment by Soliprem** [comment]
The diff in the flake.lock is entirely the alejandra formatter changing indentation

**Review by NotAShelf — CHANGES_REQUESTED** [review]
one change, lgtm otherwise

File: modules/plugins/languages/kotlin.nix [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  config = mkIf cfg.enable (mkMerge [
+    (mkIf cfg.treesitter.enable {
+      vim.treesitter.enable = true;
+      vim.treesitter.grammars = [cfg.treesitter.package];
+    })
+
+    (mkIf cfg.extraDiagnostics.enable {
+      vim.lsp.null-ls.enable = true;
+      vim.lsp.null-ls.sources = diagnosticsToLua {
+        lang = "ts";
```
[the remark this answers is not in the document]
Comment (Soliprem) [reply modules/plugins/languages/kotlin.nix:85]: absolutely a typo. Pushing the fix as soon as I test the module again

**Review by diniamo — CHANGES_REQUESTED** [review]

File: modules/plugins/languages/kotlin.nix [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    paths = [
+      pkgs.kotlin-language-server
+      pkgs.kotlin
+    ];
+    buildInputs = [pkgs.makeWrapper];
+    postBuild = ''
+      wrapProgram $out/bin/kotlin-language-server \
+        --prefix PATH : ${pkgs.lib.makeBinPath [pkgs.kotlin]}
+    '';
+  };
```
[the remark this answers is not in the document]
Comment (Soliprem) [reply modules/plugins/languages/kotlin.nix:31]: The LSP. It requires kotlin to be present at runtime (as per the logs when I simply used the lsp package).

File: modules/plugins/languages/kotlin.nix [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+          },
+          cmd = ${
+          if isList cfg.lsp.package
+          then expToLua cfg.lsp.package
+          else ''{"${cfg.lsp.package}/bin/kotlin-language-server"}''
+        },
+        }
+      '';
+    })
+  ]);
```
[the remark this answers is not in the document]
Comment (Soliprem) [reply modules/plugins/languages/kotlin.nix:110]: This follows the structure of all the other files under modules/plugins/languages. Where is the separate config file meant to be?

File: modules/plugins/languages/kotlin.nix [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  inherit (lib.types) package;
+  inherit (lib.nvim.types) mkGrammarOption diagnostics;
+  inherit (lib.lists) isList;
+  inherit (lib.nvim.lua) expToLua;
+
+  cfg = config.vim.languages.kotlin;
+
+  # Creating a version of the LSP with access to the kotlin binary.
+  # This is necessary for the LSP to load the standard library
+
```
Comment (diniamo) [inline modules/plugins/languages/kotlin.nix:20]: Remove leftover comments

File: modules/plugins/languages/kotlin.nix [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    lsp = {
+      enable = mkEnableOption "Kotlin LSP support" // {default = config.vim.languages.enableLSP;};
+
+      package = mkOption {
+        description = "kotlin_language_server package with Kotlin runtime";
+        type = package;
+        example = literalExpression ''
+          pkgs.symlinkJoin {
+              name = "Kotlin-LSP-Wrapped";
```
Comment (diniamo) [inline modules/plugins/languages/kotlin.nix:52]: Make the name lower case

File: modules/plugins/languages/kotlin.nix [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+              paths = [pkgs.kotlin-language-server];
+              nativeBuildInputs = [pkgs.makeWrapper];
+              postBuild = '''
+                wrapProgram $out/bin/kotlin-language-server \
+                  --prefix PATH : ${pkgs.lib.makeBinPath [pkgs.kotlin]}
```
Comment (diniamo) [inline modules/plugins/languages/kotlin.nix:57]: This won't render correctly, do
```suggestion
                  --prefix PATH : ''${pkgs.kotlin}/bin
```
instead

**Review by diniamo — CHANGES_REQUESTED** [review]

**Review by diniamo — APPROVED** [review]

**Pull request merged**